### PR TITLE
kubectl explain should work for both cluster and namespace resources and without a GET method

### PR DESCRIPTION
**Pull request by atiratree**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:
kubectl explain should work for both cluster and namespace resources and without a GET method

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/kubernetes/issues/118875

#### Does this PR introduce a user-facing change?

```release-note
Fixes a 1.27 regression in `kubectl explain`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl [hunk 1]
```diff
@@ -2,13 +2,33 @@
 {{- $prefix := (ternary "/api" (join "" "/apis/" $.GVR.Group) (not $.GVR.Group)) -}}
 
 {{- /* Search both cluster-scoped and namespaced-scoped paths for the GVR to find its GVK */ -}}
-{{- /* Looks for path /apis/<group>/<version>/<resource> or /apis/<group>/<version>/<version>/namespaces/{namespace}/<resource> */ -}}
+{{- /* Also search for paths with {name} component in case the list path is missing */ -}}
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl:5]: Isn't `kubectl explain bindings` going to openapiv3?, why we are only making changes on openapi/v2 templates?. I'd expect changes on v3, why am I missing?

Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl:5]: It's kinda confusing, but `explain/v2` is the v2 package of explain which is responsible for openapi v3.

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl:5]: Ah.. now it makes sense :)

**Review by alexzielenski — APPROVED** [review]
/lgtm
/approve

Inspected the new template code pretty closely and it looks like a great change. No notes. I have a PR incoming also to add an e2e to check that all resources are `explain`able so this doesn't happen again

**Comment by alexzielenski** [comment]
/cc @apelisse

for approve

**Comment by atiratree** [comment]
@alexzielenski thx for the approve and the E2Es sound like a great idea!

File: staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl [hunk 2]
```diff
@@ -2,13 +2,33 @@
...
+{{- /* Looks for the following paths: */ -}}
+{{- /* /apis/<group>/<version>/<resource> */ -}}
+{{- /* /apis/<group>/<version>/<resource>/{name} */ -}}
+{{- /* /apis/<group>/<version>/<version>/namespaces/{namespace}/<resource> */ -}}
+{{- /* /apis/<group>/<version>/<version>/namespaces/{namespace}/<resource>/{name} */ -}}
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl:10]:
```suggestion
{{- /* /apis/<group>/<version>/namespaces/{namespace}/<resource> */ -}}
{{- /* /apis/<group>/<version>/namespaces/{namespace}/<resource>/{name} */ -}}
```

is that it? Pretty sure the double-version is wrong?

File: staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext_test.go [hunk 3]
```diff
@@ -41,17 +41,46 @@ var (
 	//go:embed apiextensions.k8s.io_v1.json
 	apiextensionsJSON string
 
+	//go:embed batch.k8s.io_v1.json
+	batchJSON string
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext_test.go:45]: Any chance to find what you were looking for in one of the existing file? I'd rather not add new files if not strictly necessary (but let's take the file if there's no other choice).

File: staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl (hunk 2 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl]: good catch! fortunately it was just a comment :)

File: staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext_test.go (hunk 3 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext_test.go:45]: The difference is, that we need to test a namespaced resource and apiextensions CRDs are not namespaced. 

Also I think it is better to have a separate file for better maintenance.

```shell
$ curl localhost:6443/openapi/v3/apis/batch/v1 > batch.k8s.io_v1.json
```

This is just needed for updating the file. Merging these 2 together would be a bit harder.

**Comment by atiratree** [comment]
sig-node failures
/retest

**Comment by atiratree** [comment]
I think it is worth backporting this to 1.27. I have opened a PR with a cherry-pick: https://github.com/kubernetes/kubernetes/pull/118930

**Pull request merged**